src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java

```diff
@@ -1,77 +1,172 @@
 package io.jenkins.plugins.coverage;
 
 
+import com.google.common.collect.Sets;
+import hudson.DescriptorExtensionList;
 import hudson.FilePath;
 import hudson.model.Run;
 import hudson.model.TaskListener;
 import hudson.remoting.VirtualChannel;
-import io.jenkins.plugins.coverage.adapter.CoverageReportAdapter;
+import io.jenkins.plugins.coverage.adapter.*;
+import io.jenkins.plugins.coverage.exception.ConversionException;
 import io.jenkins.plugins.coverage.targets.CoverageElement;
 import io.jenkins.plugins.coverage.targets.CoverageResult;
+import io.jenkins.plugins.coverage.threshhold.ThreshHold;
 import jenkins.MasterToSlaveFileCallable;
+import org.apache.commons.collections.IteratorUtils;
 import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.ClassUtils;
 
 import javax.annotation.Nonnull;
 import java.io.*;
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
+import java.lang.reflect.Constructor;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class CoverageProcessor {
 
     private static final String DEFAULT_REPORT_SAVE_NAME = "coverage-report.xml";
 
-    public CoverageResult getCoverageReport(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull TaskListener listener, CoverageReportAdapter[] adapters) throws IOException, InterruptedException {
-        List<CoverageResult> results = convertToResults(run, workspace, listener, adapters);
+    private Run<?, ?> run;
+    private FilePath workspace;
+    private TaskListener listener;
+    private List<CoverageReportAdapter> adapters;
+    private List<ThreshHold> globalThreshHolds;
+
+
+    private boolean enableAutoDetect;
```

Comment:
As we discussed at the meeting, it would be great to move it into an extension point implementation if possible

Reply:
I will do this in another pull request.